### ISSUE TITLE
Loss becomes negative

### ISSUE THREAD

**Issue by ricklisz** [RobinWu218/SimDINO#12]
Hi, 

Thanks for your interesting work! I wonder if you could help with this issue: I am training SimDINOv2 using the default configuration for a custom dataset. However, I realized that some of the losses are negative from the beginning. 

The beginning of traininig is:
{"iteration": 30, "iter_time": 0.927189826965332, "data_time": 0.000664830207824707, "dino_mcr_loss": -2.606785535812378, "dino_mcr_comp_loss": 0.67333984375, "dino_mcr_global_comp_loss": 0.625, "dino_mcr_expa_loss": 1.9398465156555176, "ibot_comp_loss": -0.21865150332450867, "total_loss": -2.830775022506714, "lr": 4.5752772166451905e-06, "wd": 0.040000303773518764, "mom": 0.900000084381533, "last_layer_lr": 0, "current_batch_size": 24.0, "losses_reduced": -2.441864252090454, "total_grad_norm": 1.0972548723220825}

When the training ends, the log is:
{"iteration": 34199, "iter_time": 0.9568608999252319, "data_time": 0.003601455595344305, "dino_mcr_loss": -3.2020657062530518, "dino_mcr_comp_loss": 0.330078125, "dino_mcr_global_comp_loss": 0.931640625, "dino_mcr_expa_loss": 2.8580751419067383, "ibot_comp_loss": -0.48359355330467224, "total_loss": -3.6846115589141846, "lr": 1.0001592239302175e-06, "wd": 0.39999990810849256, "mom": 0.9999999744745812, "last_layer_lr": 1.0001592239302175e-06, "current_batch_size": 24.0, "losses_reduced": -3.234978675842285, "total_grad_norm": 0.4134863615036011}

Is this normal? Could a training log from be provided to help better understand this method? 

Thanks

**Comment by RobinWu218** [comment]
Thanks for your interest in our work! This is normal, we log the losses directly (hence they are negative). For individual values (within loss) like the coding rate term and similarity term, we log their magnitude (hence positive). Sorry this might have caused some confusion. We will streamline this in a future update. Meanwhile, you can take a look at our example log [here](https://github.com/RobinWu218/SimDINO/blob/main/assets/SimDNIOv2-vitb16.json) as reference.

**Comment by ricklisz** [comment]
Thanks for your response! I wonder if I could ask another question: I noticed that my "dino_mcr_comp_loss" eventually goes down instead of going up, however the "dino_mcr_global_comp_loss" does go up. Has that ever happened to you before? What is the purpose of tracking the "dino_mcr_global_comp_loss"?

**Comment by RobinWu218** [comment]
A higher `dino_mcr_comp_loss` usually indicates better final performance (since it means local and global features are better aligned). A decreasing `dino_mcr_comp_loss` is therefore not good. I notice that the `dino_mcr_expa_loss ` is quite a lot higher than our provided log and your `current_batch_size ` is 24. What batch size are you using and on how many gpus?

**Comment by ricklisz** [comment]
Hi, I am only using 1 GPU to debug the code as of right now so total batch is 24. I am also not running this on ImageNet but on my custom dataset (only about 9000 samples). Do you think that this could be due to the model overly focusing on _dino_mcr_expa_loss_, instead of focusing on the _dino_mcr_comp_loss_?

**Comment by RobinWu218** [comment]
The batch size might be a bit small. Note that the coding rate term computes on a covariance matrix (in this case 256 dim), and  you would get a better estimate with a larger batch size. In our experience, a batch size larger than 64 should be okay. If your gpu can only hold smaller batch size, it might be worth setting `--reduce_cov=1` to collect covariance matrices from multiple gpus via all_reduce. Regarding your last question, your intuition is correct. Other possible choices to try here include 1. increasing `--eps`, 2) reduce the gamma coefficient in `calc_expansion`. Overall, you want to make sure that `dino_mcr_comp_loss ` steadily increases over the training run.

**Comment by ricklisz** [comment]
Thanks, just one last question, do you think gradient accumulation would work for self-distillation?

**Comment by RobinWu218** [comment]
Good question -- conceptually it should work, although you need to make sure that the teacher network's update frequency is in sync with your grad acc.
